x/poolmanager/taker_fee.go

```diff
@@ -0,0 +1,182 @@
+			// If the non osmo denom is a whitelisted quote asset, we send to the community pool
+			if denomIsWhitelisted {
+				nonOsmoTakerFeeToCommunityPoolDec := takerFeeAmtRemaining.ToDec().Mul(poolManagerParams.TakerFeeParams.NonOsmoTakerFeeDistribution.CommunityPool)
+				nonOsmoTakerFeeToCommunityPoolCoins := sdk.NewCoins(sdk.NewCoin(tokenIn.Denom, nonOsmoTakerFeeToCommunityPoolDec.TruncateInt()))
+				err := k.communityPoolKeeper.FundCommunityPool(ctx, nonOsmoTakerFeeToCommunityPoolCoins, sender)
+				if err != nil {
+					return sdk.Coin{}, err
+				}
+				takerFeeAmtRemaining = takerFeeAmtRemaining.Sub(nonOsmoTakerFeeToCommunityPoolCoins.AmountOf(tokenIn.Denom))
+			} else {
+				// If the non osmo denom is not a whitelisted asset, we send to the non native fee pool for community pool module account.
+				// At epoch, this account swaps the non native, non whitelisted assets for XXX and sends to the community pool.
+				nonOsmoTakerFeeToCommunityPoolDec := takerFeeAmtRemaining.ToDec().Mul(poolManagerParams.TakerFeeParams.NonOsmoTakerFeeDistribution.CommunityPool)
+				nonOsmoTakerFeeToCommunityPoolCoins := sdk.NewCoins(sdk.NewCoin(tokenIn.Denom, nonOsmoTakerFeeToCommunityPoolDec.TruncateInt()))
+				err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, sender, nonNativeFeeCollectorForCommunityPoolName, nonOsmoTakerFeeToCommunityPoolCoins)
+				if err != nil {
+					return sdk.Coin{}, err
+				}
+				takerFeeAmtRemaining = takerFeeAmtRemaining.Sub(nonOsmoTakerFeeToCommunityPoolCoins.AmountOf(tokenIn.Denom))
+			}
+		}
+		// Staking Rewards:
+		if poolManagerParams.TakerFeeParams.NonOsmoTakerFeeDistribution.StakingRewards.GT(sdk.ZeroDec()) {
+			// Non Osmo staking rewards are sent to the non native fee pool module account
+			nonOsmoTakerFeeToStakingRewardsCoins := sdk.NewCoins(sdk.NewCoin(takerFeeCoin.Denom, takerFeeAmtRemaining))
+			err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, sender, nonNativeFeeCollectorForStakingRewardsName, nonOsmoTakerFeeToStakingRewardsCoins)
+			if err != nil {
+				return sdk.Coin{}, err
+			}
+		}
+	}
+
+	return tokenInAfterTakerFee, nil
+}
+
+func (k Keeper) calcTakerFeeExactIn(tokenIn sdk.Coin, takerFee sdk.Dec) (sdk.Coin, sdk.Coin) {
+	amountInAfterSubTakerFee := tokenIn.Amount.ToDec().Mul(sdk.OneDec().Sub(takerFee))
+	tokenInAfterSubTakerFee := sdk.NewCoin(tokenIn.Denom, amountInAfterSubTakerFee.TruncateInt())
+	takerFeeCoin := sdk.NewCoin(tokenIn.Denom, tokenIn.Amount.Sub(tokenInAfterSubTakerFee.Amount))
+
+	return tokenInAfterSubTakerFee, takerFeeCoin
+}
+
+func (k Keeper) calcTakerFeeExactOut(tokenIn sdk.Coin, takerFee sdk.Dec) (sdk.Coin, sdk.Coin) {
+	amountInAfterAddTakerFee := tokenIn.Amount.ToDec().Quo(sdk.OneDec().Sub(takerFee))
+	tokenInAfterAddTakerFee := sdk.NewCoin(tokenIn.Denom, amountInAfterAddTakerFee.RoundInt())
+	takerFeeCoin := sdk.NewCoin(tokenIn.Denom, tokenInAfterAddTakerFee.Amount.Sub(tokenIn.Amount))
+
+	return tokenInAfterAddTakerFee, takerFeeCoin
+}
```

Comment:
Why do we need this method and cannot use the one above universally?

Reply:
Hmm I think there's an important difference here. For exact out, we need to keep the output token stable, which means charging a fee should be represented as increasing the input. For exact in, we need to keep the input token stable, so we need to capture the fee by subtracting it from the input instead. I think the separate abstractions are warranted here.

Reply:
Agree with Alpin's comment, but would like to see if Alpin's comment changes Roman's view before resolving

Reply:
Agreed. Let's add comments and tests, please

Also, rounding direction should be fixed for `calcTakerFeeExactOut`:
```go
	amountInAfterAddTakerFee := tokenIn.Amount.ToDec().QuoRoundUp(sdk.OneDec().Sub(takerFee))
	tokenInAfterAddTakerFee := sdk.NewCoin(tokenIn.Denom, amountInAfterAddTakerFee.Ceil().TruncateInt())
	takerFeeCoin := sdk.NewCoin(tokenIn.Denom, tokenInAfterAddTakerFee.Amount.Sub(tokenIn.Amount))

	return tokenInAfterAddTakerFee, takerFeeCoin
```

Notice:
- `Quo` -> `QuoRoundUp`
- `RoundInt` -> `Ceil().TruncateInt()`

Reply:
I second lets add some comments pre-merge, and get rounding direction corrected

Reply:
FYI rounding direction should be fixed in my follow-up PR (#6183)